overlord/snapshotstate/backend/backend.go

```diff
@@ -323,3 +326,97 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 
 	return nil
 }
+
+type exportMetadata struct {
+	Format int       `json:"format"`
+	Date   time.Time `json:"date"`
+}
+
+// Export allows exporting a given snapshot set to the given writer.
+func Export(ctx context.Context, setID uint64, w io.Writer) error {
+	// Open all files first and keep the file descriptors open.
+	//
+	// This ensures that even if a snapshot is removed we can
+	// still serve the data. Hence no need for locking.
+	var snapshotFiles []*os.File
+	defer func() {
+		for _, f := range snapshotFiles {
+			f.Close()
+		}
+	}()
+	err := Iter(ctx, func(reader *Reader) error {
+		if reader.SetID == setID {
+			// dup() the file descriptor
+			fd, err := syscall.Dup(int(reader.Fd()))
+			if err != nil {
+				return fmt.Errorf("cannot dup %v", reader.Name())
+			}
+			f := os.NewFile(uintptr(fd), reader.Name())
+			if f == nil {
+				return fmt.Errorf("cannot handle for %v from fd %v", reader.Name(), fd)
+			}
+			snapshotFiles = append(snapshotFiles, f)
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("cannot export snapshot %v: %v", setID, err)
+	}
+	if len(snapshotFiles) == 0 {
+		return fmt.Errorf("no snapshot data found for %v", setID)
+	}
+
+	// write out a tar
+	tw := tar.NewWriter(w)
+	defer tw.Close()
+	for _, snapshotFd := range snapshotFiles {
+		stat, err := snapshotFd.Stat()
+		if err != nil {
+			return err
+		}
+		// should never happen
+		if !stat.Mode().IsRegular() {
+			return fmt.Errorf("unexported special file %q in snapshot: %s", stat.Name(), stat.Mode())
+		}
+		hdr, err := tar.FileInfoHeader(stat, "")
+		if err != nil {
+			return fmt.Errorf("symlink: %v", stat.Name())
+		}
+		if err = tw.WriteHeader(hdr); err != nil {
+			return err
+		}
+
+		// open the file
+		if _, err := io.Copy(tw, snapshotFd); err != nil {
+			return err
+		}
+	}
+
+	// write the metadata last, then the client can use that to
+	// validate the archive is complete
+	//
+	// XXX: add hashes of the individual files to the metadata?
```

Comment:
maybe we could try to at least list the files now and introduce hashes later?

Reply:
each zip contains a hash of its own, but I agree that repeating the files could be good